lib/route/cls/flower.c

```diff
@@ -598,6 +670,124 @@ int rtnl_flower_get_ip_dscp(struct rtnl_cls *cls, uint8_t *dscp, uint8_t *mask)
 	return 0;
 }
 
+/**
+ * Set IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_src(struct rtnl_cls *cls, int addr, int mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data(TC_CAST(cls))))
+		return -NLE_NOMEM;
+
+	if (addr) {
+		f->cf_ipv4_src = nl_addr_build(AF_INET,
+					       (unsigned char *)&addr,
+					       sizeof(__be32));
+
+		f->cf_mask |= FLOWER_ATTR_IPV4_SRC;
+
+		if (mask) {
+			uint32_t m = 0xFFFFFFFF << (32 - mask);
+			f->cf_ipv4_src_mask = htonl(m);
+			f->cf_mask |= FLOWER_ATTR_IPV4_SRC_MASK;
+		}
+
+		return 0;
+	}
+
+	return -NLE_FAILURE;
+}
+
+/**
+ * Get IPv4 source address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 source address
+ * @arg mask		mask for IPv4 source address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_get_ipv4_src(struct rtnl_cls *cls, struct nl_addr *addr,
+			     uint32_t *mask)
+{
+	struct rtnl_flower *f;
+
+	if (!(f = rtnl_tc_data_peek(TC_CAST(cls))))
+		return -NLE_INVAL;
+
+	if (!(f->cf_mask & FLOWER_ATTR_IPV4_SRC))
+		return -NLE_MISSING_ATTR;
+
+	if (addr)
+		memcpy(addr, f->cf_ipv4_src, AF_INET);
+
+	*mask = f->cf_ipv4_src_mask;
+
+	return 0;
+}
+
+/**
+ * Set IPv4 destination address for flower classifier
+ * @arg cls		Flower classifier.
+ * @arg addr		IPv4 destination address
+ * @arg mask		mask for IPv4 destination address
+ * @return 0 on success or a negative error code.
+ */
+int rtnl_flower_set_ipv4_dst(struct rtnl_cls *cls, int addr, int mask)
```

Comment:
`int` is not guaranteed to be a 32 bit integer (as far as C is concerned... though, I think on Linux we are guaranteed that `int` is at least 32 bits large).

also, it seems a bit ugly to use a signed integer for an IPv4 address.

We could use `uint32_t`, but I think the right choice is `in_addr_t`. The only downside of `in_addr_t`, is that we need to include `<netinet/in.h>`. Interestingly, we already use `struct in_addr` in `include/netlink/route/cls/u32.h` (is that a bug?). It seems to me that `<netinet/in.h>` is such a fundamental dependency, that we can afford to include it in libnl3's public headers (and thus have `in_addr_t`).